### PR TITLE
Use legacy transaction format and defer support of dynamic gas price

### PR DESCRIPTION
**Pull request by han0110**
This PR aims to change the transaction format to legacy one, and defer support of dynamic gas price transaction introduced by `EIP1559` to the future work for simplicity.

It also does:

- Extend `add_word` to `add_words` to match current circuit implementation
- Add `CallDataGasCost` in `tx_table` for lookup to calculate the correct intrinsic gas cost for legacy tx

### CODE REVIEW

File: src/zkevm_specs/evm/instruction.py [hunk 1]
```diff
@@ -74,32 +74,8 @@ def constrain_equal(self, lhs: int, rhs: int):
...
-
-        result, carry = self.add_word(value_with_gas_fee, sender_balance)
-        self.constrain_equal(sender_balance_prev, result)
-        self.constrain_zero(carry)
-
-        result, carry = self.add_word(value, receiver_balance_prev)
-        self.constrain_equal(receiver_balance, result)
-        self.constrain_zero(carry)
+    def constrain_sufficient_gas_left(self, gas_left: int):
+        self.int_to_bytes(gas_left, N_BYTES_GAS)
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm/instruction.py:78]: Is the intention of this function an underflow check?

Comment (han0110) [reply src/zkevm_specs/evm/instruction.py:78]: Yes

Comment (ChihChengLiang) [reply src/zkevm_specs/evm/instruction.py:78]: What do you think we rename `constrain_sufficient_gas_left ` to something like `constrain_not_underflow` to reflect the purpose?

Comment (han0110) [reply src/zkevm_specs/evm/instruction.py]: Sounds good to me. Renamed in [`6b37c0b`](https://github.com/appliedzkp/zkevm-specs/pull/78/commits/6b37c0b7f5e755c13f44642b8f04f27c13483ee9)

**Review by ChihChengLiang — APPROVED** [review]
LTGM

File: src/zkevm_specs/evm/typing.py [hunk 2]
```diff
@@ -55,8 +67,7 @@ class Transaction:
     id: int
     nonce: U64
     gas: U64
-    gas_tip_cap: U256
-    gas_fee_cap: U256
+    gas_price: U256
```
Comment (DreamWuGit) [inline src/zkevm_specs/evm/typing.py:70]: does `gas_price` really need `U256` size ?

Comment (han0110) [reply src/zkevm_specs/evm/typing.py:70]: I am not sure actually, I basically tried to follow those types used in `geth`. Some of them like `nonce` and `gas` have less bits than the one specified in yellow paper, but others match.

So I was thinking there might be some reason for `geth` to have `big.Int` for `gas_price`, so I keep it as `U256` here.

Comment (DreamWuGit) [reply src/zkevm_specs/evm/typing.py:70]: if so,  does `N_BYTES_GAS = 8` of param.py need to change accordingly ?

File: tests/evm/test_begin_tx.py [hunk 3]
```diff
@@ -42,7 +46,7 @@ def test_begin_tx(tx: Transaction, result: bool):
     block = Block()
     caller_balance_prev = rlc_store.to_rlc(int(1e20), 32)
     callee_balance_prev = rlc_store.to_rlc(0, 32)
-    caller_balance = rlc_store.to_rlc(int(1e20) - (tx.value + tx.gas * tx.gas_fee_cap), 32)
+    caller_balance = rlc_store.to_rlc(int(1e20) - (tx.value + tx.gas * tx.gas_price), 32)
     callee_balance = rlc_store.to_rlc(tx.value, 32)
 
     bytecode = Bytecode("00")
```
Comment (DreamWuGit) [inline tests/evm/test_begin_tx.py:52]: I understand for begin_tx ,the byte code `bytecode = Bytecode("00")` here doesn't matter for any input , right ?

File: src/zkevm_specs/evm/typing.py (hunk 2 above)
Comment (han0110) [reply src/zkevm_specs/evm/typing.py:70]: `N_BYTES_GAS` is for `gas` itself, tho it could have value up to 256 bits in yellow paper, but it's used as `uint64` in `geth`, so `N_BYTES_GAS = 8`.

While `gas_price` is the unit that maps `gas` to real `ether`, which also could have value up to 256 bits, and used as `big.Int` in `geth`, so we try not to change it in case there is some edge cases that haven't came to our minds.

File: tests/evm/test_begin_tx.py (hunk 3 above)
Comment (han0110) [reply tests/evm/test_begin_tx.py:52]: Here is a decision I haven't figured it out yet, thanks for reminding me, I will convert this into a issue ASAP.

The decision is, if `callee.code_hash = keccak256('')` (e.g. `callee` is not yet a contract or precompiled), and when there is call to it, we have 2 options:

1. Only do ether transfer, and not dive into another call.
2. Do ether transfer, and prepare the call context just like we are diving into another call. The first executed opcode will always be `STOP` because `program_counter == code_length` (which is not yet implemented).

The first option is the exact behavior of `geth` ([ref](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L219-L220)), but it introduces a branch if `callee.code == keccak256('')`, which requires some more constraints.

The second option is somehow weird, but it doesn't need extra constraint of `BeginTx` or `CALL` in the first option, by sacrificing an extra step, which I think is okay to do.

So the testing now have `Bytecode("00")`, which is actually having the exact behavior of the second option, to temporarily avoid this decision.

File: src/zkevm_specs/evm/typing.py (hunk 2 above)
Comment (DreamWuGit) [reply src/zkevm_specs/evm/typing.py:70]: OK to me for now! Thx

File: tests/evm/test_begin_tx.py (hunk 3 above)
Comment (DreamWuGit) [reply tests/evm/test_begin_tx.py:52]: yeah, we can think more later, I can approve this pr first for now.

**Review by DreamWuGit — APPROVED** [review]

**Review by HAOYUatHZ — APPROVED** [review]
LGTM as well!
